src/fuzzy_match.cc

```diff
@@ -578,21 +600,82 @@ namespace fuzzy
+    /* Contrastive reranking */
+    if (contrastive_factor > 0)
+    {
+      std::list<Match> candidates;
+      while (!result.empty())
+      {
+        auto match = result.top();
+        candidates.push_back(match);
+        result.pop();
+      }
+      auto comp = [contrastive_factor](const Match& m1, const Match& m2) {
+          return (m1.score - contrastive_factor * m1.penalty) < (m2.score - contrastive_factor * m2.penalty);
+      };
+      /* for memoization optimization */
+      std::unordered_map<std::pair<int, int>, float, PairHasher> edit_cost_memory;
+      while (!candidates.empty() && (number_of_matches == 0 || matches.size() < number_of_matches))
+      {
+        EditCosts internal_edit_cost = EditCosts();
+        // rescore penalties of candidates
+        for (Match &match : candidates)
+        {
+          std::vector<float> penalties;
+          for (Match &match_memory : matches)
+          {
+            auto it = edit_cost_memory.find({match.s_id, match_memory.s_id});
+            float penalty;
+            if (it != edit_cost_memory.end()) {
+              penalty = it->second;
+            } else {
+              const Costs costs(match.length, match_memory.length, internal_edit_cost);
+              penalty = _edit_distance_internal(
+                match.s, match.length,
+                match_memory.s, match_memory.length,
+                internal_edit_cost,
+                costs
+              );
+              edit_cost_memory.insert({{match.s_id, match_memory.s_id}, penalty});
+            }
+            penalty = int(10000 - penalty * 100) / 10000.0;
+            penalties.push_back(penalty);
+          }
+          if (!penalties.empty())
+          {
+            if (reduce == ContrastReduce::MAX)
+            { // max
+              match.penalty = *std::max_element(penalties.cbegin(), penalties.cend());
+            } else
+            { // mean
+              match.penalty = std::accumulate(penalties.cbegin(), penalties.cend(), 0.f) / penalties.size();
+            }
+          }
+        }
+        auto it_max = std::max_element(candidates.begin(), candidates.end(), comp);
+        matches.push_back(*it_max);
+        candidates.erase(it_max);
+        // it_max->score -= contrastive_factor * it_max->penalty;
+      }
+      return matches.size();
+    }
+    else 
     {
-      auto match = result.top();
-#ifdef DEBUG
-      std::cout << match.score << "\t" << alignment.second << "\t"
-                << match.id << std::endl;
+      while (!result.empty() && (number_of_matches == 0 || matches.size() < number_of_matches))
+      {
+        auto match = result.top();
+#ifdef XDEBUG
+        std::cout << match.score << "\t"
+                  << match.id << std::endl;
 #endif
 
-      matches.push_back(match);
+        matches.push_back(match);
 
-      result.pop();
-    }
+        result.pop();
+      }
 
-    return matches.size() > 0;
+      return matches.size() > 0;
```

Comment:
The if branch is returning `matches.size()` while the else branch is returning `matches.size() > 0`.

For consistency you can move the return statement out of the if/else block.